cjdns-splice/src/lib.rs

```diff
@@ -187,6 +187,63 @@ pub fn re_encode<L: LabelT>(
     Ok(result)
 }
 
+// todo #3
+/// This will construct a label using an array representation of a path (`path_hops`), if any label along the `path_hops` needs to be re-encoded, it will be.
+pub fn build_label<L: LabelT>(path_hops: &[PathHop<L>]) -> Result<(L, Vec<L>)> {
+    if path_hops.len() < 2 {
+        return Err(Error::NotEnoughArguments);
+    }
+
+    if path_hops.first().unwrap().label_n.is_none() || path_hops.last().unwrap().label_p.is_none() {
+        return Err(Error::ZeroLabel);
+    }
+
+    if path_hops.first().unwrap().label_p.is_some() || path_hops.last().unwrap().label_n.is_some() {
+        return Err(Error::BadArgument);
+    }
+
+    let mut ret_path = vec![path_hops.first().unwrap().label_n.unwrap()];
+    let mut ret_label = *ret_path.first().unwrap();
+
+    let hops_to_iter_over = {
+        let (_, hops_except_last) = path_hops.split_last().unwrap();
+        &hops_except_last[1..]
+    };
+
+    // Iterate over hops except for first and last
+    for hop in hops_to_iter_over {
+        if hop.label_n.is_none() || hop.label_p.is_none() {
+            return Err(Error::ZeroLabel);
+        }
+
+        // alias
+        let label_p = hop.label_p.unwrap();
+        let mut label_n = hop.label_n.unwrap();
+
+        let form_label_p = get_encoding_form(label_p, hop.encoding_scheme)?;
+        let form_label_n = get_encoding_form(label_n, hop.encoding_scheme)?;
+        if form_label_p.bit_count + form_label_p.prefix_len
+            > form_label_n.bit_count + form_label_n.prefix_len
+        {
+            label_n = re_encode(
+                label_n,
+                hop.encoding_scheme,
+                hop.encoding_scheme.get_form_idx(form_label_p),
```

Comment:
Since `get_form_idx` is a bit inefficient, lets modify `get_encoding_form` to also return the idx.